solver/src/solver/matcha_solver.rs

```diff
@@ -80,54 +81,67 @@ impl MatchaSolver {
-        let swap = match order.kind {
-            OrderKind::Sell => {
-                let query = SwapQuery {
-                    sell_token: order.sell_token,
-                    buy_token: order.buy_token,
-                    sell_amount: Some(order.sell_amount),
-                    buy_amount: None,
-                    slippage_percentage: Slippage::number_from_basis_points(
-                        STANDARD_MATCHA_SLIPPAGE_BPS,
-                    )
-                    .unwrap(),
-                    skip_validation: Some(true),
-                };
-
-                tracing::debug!("querying Matcha swap api with {:?}", query);
-                let swap = self.client.get_swap(query).await?;
-                tracing::debug!("proposed Matcha swap is {:?}", swap);
-
-                if swap.buy_amount < order.buy_amount {
-                    tracing::debug!("Order limit price not respected");
-                    return Ok(None);
+        let max_retries = 2;
+        for _ in 0..max_retries {
+            match self.try_settle_order(order.clone()).await {
+                Ok(settlement) => return Ok(settlement),
+                Err(err) if err.retryable => {
+                    tracing::debug!("Retrying Matcha settlement due to: {:?}", &err);
+                    continue;
                 }
-                swap
-            }
-            OrderKind::Buy => {
-                let query = SwapQuery {
-                    sell_token: order.sell_token,
-                    buy_token: order.buy_token,
-                    sell_amount: None,
-                    buy_amount: Some(order.buy_amount),
-                    slippage_percentage: Slippage::number_from_basis_points(
-                        STANDARD_MATCHA_SLIPPAGE_BPS,
-                    )
-                    .unwrap(),
-                    skip_validation: Some(true),
-                };
-
-                tracing::debug!("querying Matcha swap api with {:?}", query);
-                let swap = self.client.get_swap(query).await?;
-                tracing::debug!("proposed Matcha swap is {:?}", swap);
-
-                if swap.sell_amount > order.sell_amount {
-                    tracing::debug!("Order limit price not respected");
-                    return Ok(None);
-                }
-                swap
+                Err(err) => return Err(err.inner),
             }
+        }
+        // One last attempt, else throw converted error
+        self.try_settle_order(order).await.map_err(|err| err.inner)
+    }
+
+    fn account(&self) -> &Account {
+        &self.account
+    }
+
+    fn name(&self) -> &'static str {
+        "Matcha"
+    }
+}
+
+impl From<MatchaResponseError> for SettlementError {
+    fn from(err: MatchaResponseError) -> Self {
+        SettlementError {
+            inner: anyhow!("Matcha Response Error {:?}", err),
+            retryable: matches!(err, MatchaResponseError::ServerError(_)),
```

Comment:
Maybe try_settle_order could return MatchaResponseError and we have a function here is `fn is_retryable(error: &MatchaResponseError) -> bool`. I don't think the intermediate SettlementError type is that useful.

Reply:
I was mostly doing this to be consistent with the pattern we have for Paraswap solver. I see that is not super important to have here, but just thought that it might become more relevant to do things this way once we encounter more than just one retryable matcha error. 

I will revert this part for now until it is needed. Would you also suggest that this `SettlementError` isn't needed on the Paraswap side either?

Reply:
It is no big deal either way. It just felt like there no advantage in storing it in a struct when a plain function (or could be implemented on MatchaResponseError directly) suffices.
If the pattern is the same for Paraswap then we could make the same change there too.

Reply:
Actually the reason for this goes a bit beyond just the retryable field. We also use this so that the ? operator can be used within try_settle_order. I think it is probably best to leave it as it is.